File: android/lca/src/main/java/com/dreampany/lca/data/model/Coin.java
/*
package com.dreampany.lca.data.model;

import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;

import com.dreampany.frame.data.model.Base;
import com.dreampany.lca.data.enums.CoinSource;
import com.dreampany.lca.data.enums.Currency;
import com.dreampany.lca.misc.Constants;
import com.google.common.collect.Maps;
import com.google.firebase.database.Exclude;
import com.google.firebase.database.IgnoreExtraProperties;
import com.google.firebase.database.PropertyName;

import org.apache.commons.lang3.builder.ReflectionToStringBuilder;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

*/
/**
 * Created by Hawladar Roman on 29/5/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 *//*

@Entity(indices = {@Index(value = {Constants.Coin.ID}, unique = true)},
        primaryKeys = {Constants.Coin.ID})
@IgnoreExtraProperties
public class Coin extends Base {

    private CoinSource source;
    private String name;
    private String symbol;
    private String slug;
    private int rank;
    @PropertyName(Constants.Coin.MARKET_PAIRS)
    private int marketPairs;
    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    private double circulatingSupply;
    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    private double totalSupply;
    @PropertyName(Constants.Coin.MAX_SUPPLY)
    private double maxSupply;
    @PropertyName(Constants.Coin.LAST_UPDATED)
    private long lastUpdated;
    @PropertyName(Constants.Coin.DATE_ADDED)
    private long dateAdded;
    private List<String> tags;
    @Ignore
    @Exclude
    private Map<Currency, Quote> quotes;

    @Ignore
    public Coin() {

    }

    public Coin(@NonNull String id) {
        this.id = id;
    }

    @Ignore
    private Coin(Parcel in) {
        super(in);
        source = in.readParcelable(CoinSource.class.getClassLoader());
        name = in.readString();
        symbol = in.readString();
        slug = in.readString();
        rank = in.readInt();
        marketPairs = in.readInt();
        circulatingSupply = in.readDouble();
        totalSupply = in.readDouble();
        maxSupply = in.readDouble();
        lastUpdated = in.readLong();
        dateAdded = in.readLong();
        tags = in.createStringArrayList();
        quotes = (Map<Currency, Quote>) in.readSerializable();
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeParcelable(source, flags);
        dest.writeString(name);
        dest.writeString(symbol);
        dest.writeString(slug);
        dest.writeInt(rank);
        dest.writeInt(marketPairs);
        dest.writeDouble(circulatingSupply);
        dest.writeDouble(totalSupply);
        dest.writeDouble(maxSupply);
        dest.writeLong(lastUpdated);
        dest.writeLong(dateAdded);
        dest.writeStringList(tags);
        dest.writeSerializable((Serializable) quotes);
    }

    public static final Creator<Coin> CREATOR = new Creator<Coin>() {
        @Override
        public Coin createFromParcel(Parcel in) {
            return new Coin(in);
        }

        @Override
        public Coin[] newArray(int size) {
            return new Coin[size];
        }
    };

    @NonNull
    @Override
    public String toString() {
        return ReflectionToStringBuilder.toString(this);
    }

    public void setSource(CoinSource source) {
        this.source = source;
    }

    public CoinSource getSource() {
        return source;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setSymbol(String symbol) {
        this.symbol = symbol;
    }

    public String getSymbol() {
        return symbol;
    }

    public void setSlug(String slug) {
        this.slug = slug;
    }

    public String getSlug() {
        return slug;
    }

    public void setRank(int rank) {
        this.rank = rank;
    }

    public int getRank() {
        return rank;
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    public void setMarketPairs(int marketPairs) {
        this.marketPairs = marketPairs;
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    public int getMarketPairs() {
        return marketPairs;
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    public void setCirculatingSupply(double circulatingSupply) {
        this.circulatingSupply = circulatingSupply;
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    public double getCirculatingSupply() {
        return circulatingSupply;
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    public void setTotalSupply(double totalSupply) {
        this.totalSupply = totalSupply;
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    public double getTotalSupply() {
        return totalSupply;
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    public void setMaxSupply(double maxSupply) {
        this.maxSupply = maxSupply;
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    public double getMaxSupply() {
        return maxSupply;
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    public void setLastUpdated(long lastUpdated) {
        this.lastUpdated = lastUpdated;
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    public long getLastUpdated() {
        return lastUpdated;
    }

    @Exclude
    public Date getLastUpdatedDate() {
        return new Date(getLastUpdated());
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    public void setDateAdded(long dateAdded) {
        this.dateAdded = dateAdded;
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    public long getDateAdded() {
        return dateAdded;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setQuotes(Map<Currency, Quote> quotes) {
        this.quotes = quotes;
    }

    public void addQuote(Quote quote) {
        if (quotes == null) {
            quotes = Maps.newHashMap();
        }
        quotes.put(quote.getCurrency(), quote);
    }

    @Exclude
    public Map<Currency, Quote> getQuotes() {
        return quotes;
    }

    @Exclude
    public List<Quote> getQuotesAsList() {
        if (quotes == null) {
            return null;
        }
        return new ArrayList<>(quotes.values());
    }

    public void clearQuote() {
        quotes.clear();
    }

    public boolean hasQuote() {
        if (quotes == null) {
            return false;
        }
        return !quotes.isEmpty();
    }

    public boolean hasQuote(String currency) {
        if (quotes == null) {
            return false;
        }
        return quotes.containsKey(Currency.valueOf(currency));
    }

    public boolean hasQuote(Currency currency) {
        if (quotes == null) {
            return false;
        }
        return quotes.containsKey(currency);
    }

    public boolean hasQuote(Currency[] currencies) {
        if (quotes == null) {
            return false;
        }
        for (Currency currency : currencies) {
            if (!quotes.containsKey(currency)) {
                return false;
            }
        }
        return true;
    }

    public void addQuotes(List<Quote> quotes) {
        for (Quote quote : quotes) {
            addQuote(quote);
        }
    }

    public Quote getQuote(Currency currency) {
        if (quotes != null) {
            return quotes.get(currency);
        }
        return null;
    }

    @Exclude
    public Quote getLatestQuote() {
        Quote latest = null;
        if (quotes != null) {
            for (Map.Entry<Currency, Quote> entry : quotes.entrySet()) {
                if (latest == null || latest.getTime() < entry.getValue().getTime()) {
                    latest = entry.getValue();
                }
            }
        }
        return latest;
    }
}
*/

File: android/lca/src/main/kotlin/com/dreampany/lca/data/model/Coin.kt
package com.dreampany.lca.data.model

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.dreampany.framework.util.TimeUtilKt
import com.dreampany.framework.data.model.Base
import com.dreampany.lca.data.enums.CoinSource
import com.dreampany.lca.data.enums.Currency
import com.dreampany.lca.misc.Constants
import com.google.common.collect.Maps
import com.google.firebase.database.Exclude
import com.google.firebase.database.IgnoreExtraProperties
import com.google.firebase.database.PropertyName
import kotlinx.android.parcel.Parcelize
import java.util.*

/**
 * Created by roman on 2019-07-28
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Parcelize
@IgnoreExtraProperties
@Entity(
    indices = [Index(
        value = [Constants.Coin.ID],
        unique = true
    )],
    primaryKeys = [Constants.Coin.ID]
)
data class Coin(
    override var time: Long = Constants.Default.LONG,
    override var id: String = Constants.Default.STRING,
    var source: CoinSource? = Constants.Default.NULL,
    var name: String? = Constants.Default.NULL,
    var symbol: String? = Constants.Default.NULL,
    var slug: String? = Constants.Default.NULL,
    var rank: Int = Constants.Default.INT,

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    private var marketPairs: Int = Constants.Default.INT,

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    private var circulatingSupply: Double = Constants.Default.DOUBLE,

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    private var totalSupply: Double = Constants.Default.DOUBLE,

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    private var maxSupply: Double = Constants.Default.DOUBLE,

    @PropertyName(Constants.Coin.LAST_UPDATED)
    private var lastUpdated: Long = Constants.Default.LONG,

    @PropertyName(Constants.Coin.DATE_ADDED)
    private var dateAdded: Long = Constants.Default.LONG,

    var tags: MutableList<String>? = Constants.Default.NULL,

    @Ignore
    @Exclude
    private var quotes: MutableMap<Currency, Quote>? =  Constants.Default.NULL

) : Base() {

    @Ignore
    constructor() : this(time = TimeUtilKt.currentMillis()) {}

    constructor(id: String) : this(time = TimeUtilKt.currentMillis(), id = id) {}

/*    @Ignore
    private constructor (parcel: Parcel) : this(parcel.readLong(), parcel.readString()!!) {
        source = parcel.readParcelable(CoinSource::class.java.classLoader)
        name = parcel.readString()
        symbol = parcel.readString()
        slug = parcel.readString()
        rank = parcel.readInt()
        marketPairs = parcel.readInt()
        circulatingSupply = parcel.readDouble()
        totalSupply = parcel.readDouble()
        maxSupply = parcel.readDouble()
        lastUpdated = parcel.readLong()
        dateAdded = parcel.readLong()
        tags = parcel.createStringArrayList()
        quotes = parcel.readSerializable() as MutableMap<Currency, Quote>?

    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(time)
        dest.writeString(id)
        dest.writeParcelable(source, flags)
        dest.writeString(name)
        dest.writeString(symbol)
        dest.writeString(slug)
        dest.writeInt(rank)
        dest.writeInt(marketPairs)
        dest.writeDouble(circulatingSupply)
        dest.writeDouble(totalSupply)
        dest.writeDouble(maxSupply)
        dest.writeLong(lastUpdated)
        dest.writeLong(dateAdded)
        dest.writeStringList(tags)
        dest.writeSerializable(quotes as Serializable)
    }*/

/*    companion object CREATOR : Parcelable.Creator<Coin> {
        override fun createFromParcel(parcel: Parcel): Coin {
            return Coin(parcel)
        }

        override fun newArray(size: Int): Array<Coin?> {
            return arrayOfNulls(size)
        }
    }*/

    override fun toString(): String {
        return "Coin ($id) == $id"
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    fun setMarketPairs(marketPairs: Int) {
        this.marketPairs = marketPairs
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    fun getMarketPairs(): Int {
        return marketPairs
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    fun setCirculatingSupply(circulatingSupply: Double) {
        this.circulatingSupply = circulatingSupply
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    fun getCirculatingSupply(): Double {
        return circulatingSupply
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    fun setTotalSupply(totalSupply: Double) {
        this.totalSupply = totalSupply
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    fun getTotalSupply(): Double {
        return totalSupply
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    fun setMaxSupply(maxSupply: Double) {
        this.maxSupply = maxSupply
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    fun getMaxSupply(): Double {
        return maxSupply
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    fun setLastUpdated(lastUpdated: Long) {
        this.lastUpdated = lastUpdated
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    fun getLastUpdated(): Long {
        return lastUpdated
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    fun setDateAdded(dateAdded: Long) {
        this.dateAdded = dateAdded
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    fun getDateAdded(): Long {
        return dateAdded
    }

    @Exclude
    fun getLastUpdatedDate(): Date {
        return Date(getLastUpdated())
    }

    fun addQuote(quote: Quote) {
        if (quotes == null) {
            quotes = Maps.newHashMap()
        }
        quotes!![quote.currency] = quote
    }

    @Exclude
    fun getQuotes(): Map<Currency, Quote>? {
        return quotes
    }

    @Exclude
    fun getQuotesAsList(): List<Quote>? {
        return if (quotes == null) {
            null
        } else ArrayList(quotes!!.values)
    }

    fun clearQuote() {
        quotes?.clear()
    }

    fun hasQuote(): Boolean {
        return !quotes.isNullOrEmpty()
    }

    fun hasQuote(currency: String): Boolean {
        return if (quotes == null) {
            false
        } else quotes!!.containsKey(Currency.valueOf(currency))
    }

    fun hasQuote(currency: Currency): Boolean {
        return if (quotes == null) {
            false
        } else quotes!!.containsKey(currency)
    }

    fun hasQuote(currencies: Array<Currency>): Boolean {
        if (quotes == null) {
            return false
        }
        for (currency in currencies) {
            if (!quotes!!.containsKey(currency)) {
                return false
            }
        }
        return true
    }

    fun addQuotes(quotes: List<Quote>) {
        for (quote in quotes) {
            addQuote(quote)
        }
    }

    fun getQuote(currency: Currency): Quote? {
        return if (quotes != null) {
            quotes!!.get(currency)
        } else null
    }

    @Exclude
    fun getLatestQuote(): Quote? {
        var latest: Quote? = null
        quotes?.forEach {
            if (latest == null || latest!!.time < it.value.time) {
                latest = it.value
            }
        }
        return latest
    }
}
